src/game_api/spawn_api.cpp

```diff
@@ -662,18 +662,33 @@ void init_spawn_hooks()
     }
 }
 
-void spawn_player(int8_t player_slot, float x, float y)
+int32_t spawn_player(int8_t player_slot, std::optional<float> x, std::optional<float> y)
 {
     if (player_slot < 1 || player_slot > 4)
-        return;
+        return -1;
+    auto state = State::get().ptr();
+    auto& slot = state->items->player_select_slots[player_slot - 1];
+    if (slot.character < to_id("ENT_TYPE_CHAR_ANA_SPELUNKY") || slot.character > to_id("ENT_TYPE_CHAR_CLASSIC_GUY"))
+        return -1;
+    if (state->items->player_count < player_slot)
+        state->items->player_count = player_slot;
+    slot.activated = true;
 
     push_spawn_type_flags(SPAWN_TYPE_SCRIPT);
     OnScopeExit pop{[]
                     { pop_spawn_type_flags(SPAWN_TYPE_SCRIPT); }};
 
-    using spawn_player_fun = void(Items*, uint8_t ps, float pos_x, float pos_y);
+    auto old_x = state->level_gen->spawn_x;
+    auto old_y = state->level_gen->spawn_y;
+    state->level_gen->spawn_x = x.value_or(old_x);
+    state->level_gen->spawn_y = y.value_or(old_y);
+    auto uid = (int32_t)state->next_entity_uid;
+    using spawn_player_fun = void(Items*, uint8_t ps);
     static auto spawn_player = (spawn_player_fun*)get_address("spawn_player");
-    spawn_player(get_state_ptr()->items, player_slot - 1, x, y);
+    spawn_player(get_state_ptr()->items, player_slot - 1);
```

Comment:
I'm pretty sure it didn't even have those params and all players just spawned at the level_gen->spawn coordinates anyway.

Reply:
Yes, it only takes 2 parameters, is mostly called from the virtual populate_players in theme
For the layer hack, it doesn't cover the HH spawn, player ghost and much more (powerups maybe?)

Reply:
What do those things have to do with spawning a player in the back layer manually?

Reply:
The function does not spawn player only, it spawn all the stuff from the inventory, so all the power ups, all the hired hand that were assigned to the player the previews levels etc.

Reply:
Oh right, I didn't really read it through. Not very applicable to the scenario I was thinking about, but it would probably be a problem in let's say some interconnected world where you can enter and leave the level from back layer with some items too, not just debugging in OL

Reply:
Function is in the API, if it was only for debugging/UI I would not be concerned
It's probably possible to make the layer stuff work, but it needs more code, probably the easiest would be to look thru it and change all the `0x1300` patterns, + the HH probably needs the `patch_and_redirect` as there is not enough space to set the layer param for that function

Reply:
ok hear me out

Reply:
That's so stupid, i love it xd
thou you probably should not use ` write_mem_recoverable` as it's saves the old memory once, so if the pointer changes it will recover to the wrong address

Reply:
Also no need to hardcode `0x1300` offset in this case, you can just change the layer pointers around

Reply:
Yeah I don't know why I wrote it like that